src/main/java/org/commcare/formplayer/services/FormDefinitionService.java

```diff
@@ -54,4 +54,24 @@ public SerializableFormDefinition getOrCreateFormDefinition(
             return this.formDefinitionRepo.save(newFormDef);
         });
     }
+
+    /**
+     * This method uses the sessionId to cache deserialized FormDefs
+     * Deserializing a serialized FormDef object is costly, and this avoids doing so in between requests
+     * within the same session
+     *
+     * @param sessionId form session id
+     * @param serializedFormDef serialized FormDef object
+     * @return deserialized FormDef object
+     */
+    @Cacheable(key = "#sessionId")
+    public FormDef getFormDef(String sessionId, String serializedFormDef) {
+        FormDef formDef;
+        try {
+            formDef = FormDefStringSerializer.deserialize(serializedFormDef);
```

Comment:
You could simplify this for callers by taking in the whole session object instead of passing the ID and serialized formDef separately:

```suggestion
    @Cacheable(key = "#session.id")
    public FormDef getFormDef(SerializableFormSession session) {
        FormDef formDef;
        try {
            formDef = FormDefStringSerializer.deserialize(session.getFormDefinition().getSerializedFormDef());
```

Reply:
Good point. See 5068f204cea26929ec844d477ed79d290fb0567c